src/components/tl/mlx5/tl_mlx5_team.c

```diff
@@ -169,40 +146,169 @@ static inline ucc_status_t ucc_tl_mlx5_a2a_team_test(ucc_base_team_t *team)
+                 * mcast team */
+                tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                         "failure during mcast ctx create, no mcast team support");
+
+                if (tl_team->local_mcast_ctx_ready) {
+                    comm = tl_team->mcast->mcast_comm;
+                    /* release the resources */
+                    if (ibv_dereg_mr(comm->grh_mr)) {
+                        tl_warn(UCC_TL_TEAM_LIB(tl_team),
+                                "ibv_dereg_mr failed");
+                    }
+                    if (ibv_destroy_cq(comm->rcq)) {
+                        tl_warn(UCC_TL_TEAM_LIB(tl_team),
+                                "ibv_destroy_cq failed");
+                    }
+                    ucc_free(comm->params.oob);
+                    ucc_free(comm);
+                    ucc_free(tl_team->mcast);
+                }
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
+            } else {
+                tl_debug(UCC_TL_TEAM_LIB(tl_team),
+                         "all team members have mcast ctx ready");
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_INIT;
+            }
+
+            return UCC_INPROGRESS;
+        } else {
+            if (tl_team->global_status_array[0] != UCC_OK) {
+                //a2a team not avail for some of nodes so disable it
+                if (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY) {
+                    // free the resources
+                    ucc_tl_mlx5_alltoall_cleanup(tl_team);
+                }
+                tl_team->a2a_state = TL_MLX5_TEAM_STATE_ALLTOALL_NOT_AVAILABLE;
+            }
+
+            if (tl_team->global_status_array[1] != UCC_OK) {
+                //mcast team not avail for some of nodes so disable it
+                if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY) {
+                    // free the resources
+                    ucc_tl_mlx5_clean_mcast_comm(tl_team->mcast->mcast_comm);
+                }
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_NOT_AVAILABLE;
+            }
+
+            tl_debug(team->context->lib, "initialized tl team: %p: MCAST component is %s ALLTOALL component is %s",
+                    team, (tl_team->mcast_state == TL_MLX5_TEAM_STATE_MCAST_READY)?"ENABLED":"DISABLED",
+                    (tl_team->a2a_state == TL_MLX5_TEAM_STATE_ALLTOALL_READY)?"ENABLED":"DISABLED");
+        }
+
+        ucc_free(tl_team->local_status_array);
+        ucc_free(tl_team->global_status_array);
+
+        return UCC_OK;
+    }
+
+    ucc_assert(!tl_team->global_sync_in_progress);
+    
+    if (tl_team->mcast_state == TL_MLX5_TEAM_STATE_INIT &&
+            tl_team->a2a_state == TL_MLX5_TEAM_STATE_INIT ) {
+        // check if ctx is ready for a2a and mcast
+        tl_team->local_status_array  = ucc_calloc(1, 2*sizeof(ucc_status_t));
```

Comment:
those two arrays should be allocated statically. I would also suggest that indexing into it (occuring in many places in this file) should be made with macros or enum type instead of hardcoded, like
```
#define TL_MLX5_A2A_STATUS_INDEX 0
#define TL_MLX5_MCAST_STATUS_INDEX 1
```